pkg/tasks/delete_snapshots.go

```diff
@@ -0,0 +1,243 @@
+	"github.com/content-services/content-sources-backend/pkg/dao"
+	"github.com/content-services/content-sources-backend/pkg/db"
+	"github.com/content-services/content-sources-backend/pkg/models"
+	"github.com/content-services/content-sources-backend/pkg/pulp_client"
+	"github.com/content-services/content-sources-backend/pkg/tasks/helpers"
+	"github.com/content-services/content-sources-backend/pkg/tasks/payloads"
+	"github.com/content-services/content-sources-backend/pkg/tasks/queue"
+	"github.com/content-services/content-sources-backend/pkg/utils"
+)
+
+type DeleteSnapshots struct {
+	orgID          string
+	rhDomainName   string
+	domainName     string
+	ctx            context.Context
+	payload        *payloads.DeleteSnapshotsPayload
+	task           *models.TaskInfo
+	daoReg         *dao.DaoRegistry
+	pulpClient     *pulp_client.PulpClient
+	pulpDistHelper *helpers.PulpDistributionHelper
+}
+
+func DeleteSnapshotsHandler(ctx context.Context, task *models.TaskInfo, _ *queue.Queue) error {
+	opts := payloads.DeleteSnapshotsPayload{}
+	if err := json.Unmarshal(task.Payload, &opts); err != nil {
+		return fmt.Errorf("payload incorrect type for %s", config.DeleteSnapshotsTask)
+	}
+
+	logger := LogForTask(task.Id.String(), task.Typename, task.RequestID)
+	ctxWithLogger := logger.WithContext(ctx)
+
+	daoReg := dao.GetDaoRegistry(db.DB)
+	domainName, err := daoReg.Domain.Fetch(ctxWithLogger, task.OrgId)
+	if err != nil {
+		return err
+	}
+
+	rhDomainName, err := daoReg.Domain.Fetch(ctxWithLogger, config.RedHatOrg)
+	if err != nil {
+		return err
+	}
+
+	pulpClient := pulp_client.GetPulpClientWithDomain(domainName)
+
+	ds := DeleteSnapshots{
+		orgID:        task.OrgId,
+		rhDomainName: rhDomainName,
+		domainName:   domainName,
+		ctx:          ctx,
+		payload:      &opts,
+		task:         task,
+		daoReg:       daoReg,
+		pulpClient:   &pulpClient,
+	}
+	return ds.Run()
+}
+
+func (ds *DeleteSnapshots) Run() error {
+	if config.PulpConfigured() && ds.pulpClient != nil {
+		err := ds.configurePulpClient()
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, snapUUID := range ds.payload.SnapshotsUUIDs {
+		templateUpdateMap := make(map[string]models.Snapshot)
+		snap, err := ds.daoReg.Snapshot.FetchUnscoped(ds.ctx, snapUUID)
+		if err != nil {
+			return err
+		}
+		repo, err := ds.daoReg.RepositoryConfig.Fetch(ds.ctx, ds.orgID, snap.RepositoryConfigurationUUID)
+		if err != nil {
+			return err
+		}
+
+		err = ds.updateTemplatesUsingSnap(&templateUpdateMap, snap)
+		if err != nil {
+			return err
+		}
+		err = ds.daoReg.Template.DeleteTemplateSnapshot(ds.ctx, snap.UUID)
+		if err != nil {
+			return err
+		}
+		if repo.LastSnapshotUUID == snapUUID {
+			err = ds.updateRepoConfig(repo.UUID, snapUUID)
+			if err != nil {
+				return err
+			}
+		}
+
+		if config.PulpConfigured() && ds.pulpClient != nil {
+			err = ds.deleteOrUpdatePulpContent(snap, repo, templateUpdateMap)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
pulp configuration is checked at the beginning of the task, so I would remove the check here

Reply:
pulp configuration is checked at the beginning to only setup/configure the the pulp client, it doesn't return if it ain't configured and as the `deleteOrUpdatePulpContent()` function requires pulp, I think the check should remain here 😅

Reply:
ah okay, you're right I see that distinction now. 

In that case, maybe we should return an error if pulp isn't configured right at the beginning? If it's not configured, and we delete some things while only skipping pulp, I worry that would cause some weird behavior or leave objects in pulp we can't delete anymore.

Reply:
yeah, that makes sense, but I don't know if that's what we want to do 🤔 
I took this approach, because in the `delete_repository_snapshots` task it's also done this way, the pulp objects are deleted only if pulp is configured and the repo config is deleted even if not 😅 
maybe we should evaluate if this is correct in both of these tasks in another ticket and change this in both if we don't wanna do it this way? 💭